src/Symfony/Component/Validator/Mapping/Loader/PropertyInfoLoader.php

```diff
@@ -0,0 +1,108 @@
+use Symfony\Component\Validator\Constraints\NotNull;
+use Symfony\Component\Validator\Constraints\Type;
+use Symfony\Component\Validator\Mapping\AutoMappingTrait;
+use Symfony\Component\Validator\Mapping\ClassMetadata;
+
+/**
+ * Guesses and loads the appropriate constraints using PropertyInfo.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class PropertyInfoLoader implements LoaderInterface
+{
+    use AutoMappingTrait;
+
+    private $listExtractor;
+    private $typeExtractor;
+    private $namespaces;
+
+    public function __construct(PropertyListExtractorInterface $listExtractor, PropertyTypeExtractorInterface $typeExtractor, array $namespaces = null)
+    {
+        $this->listExtractor = $listExtractor;
+        $this->typeExtractor = $typeExtractor;
+        $this->namespaces = $namespaces;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadClassMetadata(ClassMetadata $metadata)
+    {
+        $className = $metadata->getClassName();
+        if (!$this->isClassAutoMapped($className, $this->namespaces)) {
+            return false;
+        }
+
+        if (!$properties = $this->listExtractor->getProperties($className)) {
+            return false;
+        }
+
+        foreach ($properties as $property) {
+            $types = $this->typeExtractor->getTypes($metadata->getClassName(), $property);
+            if (null === $types) {
+                continue;
+            }
+
+            $hasTypeConstraint = false;
+            $hasNotNullConstraint = false;
+            $hasNotBlankConstraint = false;
+            foreach ($metadata->getPropertyMetadata($property) as $propertyMetadata) {
+                foreach ($propertyMetadata->getConstraints() as $constraint) {
+                    if ($constraint instanceof Type) {
+                        $hasTypeConstraint = true;
+                    } elseif ($constraint instanceof NotNull) {
+                        $hasNotNullConstraint = true;
+                    } elseif ($constraint instanceof NotBlank) {
+                        $hasNotBlankConstraint = true;
+                    }
+                }
+            }
+
+            $builtinTypes = array();
+            $nullable = false;
+            $scalar = true;
+
+            foreach ($types as $type) {
+                $builtinTypes[] = $type->getBuiltinType();
+
+                if ($scalar && !\in_array($type->getBuiltinType(), array(PropertyInfoType::BUILTIN_TYPE_INT, PropertyInfoType::BUILTIN_TYPE_FLOAT, PropertyInfoType::BUILTIN_TYPE_STRING, PropertyInfoType::BUILTIN_TYPE_BOOL), true)) {
+                    $scalar = false;
+                }
+
+                if (!$nullable && $type->isNullable()) {
+                    $nullable = true;
+                }
+            }
+            if (!$hasTypeConstraint) {
+                if (1 === \count($builtinTypes)) {
+                    $metadata->addPropertyConstraint($property, new Type(array('type' => $builtinTypes[0])));
```

Comment:
If the builtin type is object, we should try adding a `Type` constraint specific to the `PropertyInfoType::getClassName()` returned FQCN.

If it's a collection, we should be able to add a `@All({ Type(collectionValueType) })` constraint.